x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/JobTaskState.java

```diff
@@ -54,28 +66,46 @@ public static JobTaskState fromXContent(XContentParser parser) {
     private final JobState state;
     private final long allocationId;
     private final String reason;
+    private final Instant lastStateChangeTime;
 
-    public JobTaskState(JobState state, long allocationId, @Nullable String reason) {
+    public JobTaskState(JobState state, long allocationId, @Nullable String reason, @Nullable Instant lastStateChangeTime) {
         this.state = Objects.requireNonNull(state);
         this.allocationId = allocationId;
         this.reason = reason;
+        // Round to millisecond to avoid serialization round trip differences
+        this.lastStateChangeTime = (lastStateChangeTime != null) ? Instant.ofEpochMilli(lastStateChangeTime.toEpochMilli()) : null;
```

Comment:
Just curious how a round trip affects the serialization referred to here?

Reply:
On [line 144](https://github.com/elastic/elasticsearch/pull/102231/files#diff-5d64e65f29b1de420eedf4383bd5e5b3c46d13231a17ad2f265775a8188542a0R144) with write the timestamp to X-Content with millisecond accuracy. That's standard across all places in Elasticsearch that write timestamps to X-Content. So a problem arises if we write an object to X-Content that contains a timestamp that has non-zero nanoseconds that aren't an exact millisecond, then parse a new object back from X-Content and assert that it's equal to the original one. This is what the serialization test framework does.

You'll see the pattern of rounding timestamps to millisecond accuracy in many places across the ES codebase.

Reply:
Ah gotcha, thanks!